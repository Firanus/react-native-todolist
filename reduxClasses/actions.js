export const types = {
  ADD_ITEM: 'ADD_ITEM',
  REMOVE_COMPLETED_ITEMS: 'REMOVE_COMPLETED_ITEMS',
  REMOVE_ITEM: 'REMOVE_ITEM',
  TOGGLE_COMPLETED: 'TOGGLE_COMPLETED'
};

export const actionCreators = {
  add: (itemTitle) => {
    return {type: types.ADD_ITEM, payload: itemTitle}
  },
  removeCompleted: () => {
    return {type: types.REMOVE_COMPLETED_ITEMS}
  }
};