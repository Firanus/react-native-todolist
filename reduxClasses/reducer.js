import {types} from './actions'

const initialState = {
  todos: [ {id: 0, title: 'defaultItem', isCompleted: false} ]
};

export const reducer = (state = initialState, action) => {
  const { todos } = state;
  const { type, payload } = action;

  switch(type){
    case types.ADD_ITEM:
      return {
        ...state,
        todos: [
          {
            id: state.todos.reduce((maxId, elem) => elem.id > maxId ? elem.id : maxId, 0) + 1,
            title: payload,
            isCompleted: false
          },
          ...state.todos],
      };
    case types.REMOVE_COMPLETED_ITEMS:
      return {
        ...state,
        todos: state.todos.filter((todo) => {
          return todo.isCompleted = false
        }),
      };
      break;
    default:
      break;
  }

  return state;
};