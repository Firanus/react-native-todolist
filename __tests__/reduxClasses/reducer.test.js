import { reducer } from '../../reduxClasses/reducer'
import { types } from '../../reduxClasses/actions'


describe('The reducer', () => {
  test('Adds an item', () => {
    const initialState = {
      todos: [ {id: 0, title: 'Default Item', isCompleted: false} ]
    };
    const addActionMock = {
      type: types.ADD_ITEM,
      payload: 'My Added Todo'
    };

    expect(reducer(initialState, addActionMock)).toMatchSnapshot();
  });

  test('Removes completed items', () => {
    const initialState = {
      todos: [
        {id: 0, title: 'Default Item', isCompleted: false},
        {id: 1, title: 'First item to remove', isCompleted: true},
        {id: 2, title: 'Second item to remove', isCompleted: true}
      ]
    };
    const removeCompletedActionMock = {
      type: types.REMOVE_COMPLETED_ITEMS
    };

    expect(reducer(initialState, removeCompletedActionMock)).toMatchSnapshot();
  })
});