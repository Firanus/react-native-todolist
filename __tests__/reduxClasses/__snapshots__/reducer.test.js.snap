// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The reducer Adds an item 1`] = `
Object {
  "todos": Array [
    Object {
      "id": 1,
      "isCompleted": false,
      "title": "My Added Todo",
    },
    Object {
      "id": 0,
      "isCompleted": false,
      "title": "Default Item",
    },
  ],
}
`;

exports[`The reducer Removes completed items 1`] = `
Object {
  "todos": Array [
    Object {
      "id": 0,
      "isCompleted": false,
      "title": "Default Item",
    },
  ],
}
`;
